packages/flutter/lib/src/painting/debug.dart

```diff
@@ -31,6 +31,24 @@ typedef HttpClientProvider = HttpClient Function();
 /// This value is ignored in non-debug builds.
 HttpClientProvider debugNetworkImageHttpClientProvider;
 
+
+/// If this value is a positive number, the framework will check that images
+/// it renders are not more than that many kilobytes greater than the minimum
+/// size needed to display the image at its current resolution without scaling.
+///
+/// For example, if a 100x100 image is decoded it takes roughly 53kb in memory
+/// (including mipmapping overhead). If it is only ever displayed at 50x50, it
+/// would take only 13kb if the cacheHeight/cacheWidth parameters had been
+/// specified at that size. This problem becomes more serious for larger
+/// images, such as a high resolution image from a 12MP camera, which would be
+/// 64mb when decoded.
+///
+/// When using this setting, developers should consider whether the image will
+/// be panned or scaled up in the application, how many images are being
+/// displayed, and whether the application will run on multiple devices with
+/// different resolutions and memory capacities.
+int debugImageOverheadAllowedInKilobytes;
```

Comment:
should there also be a percentage threshold for this as well where we only warn if both the KB and percentage thresholds are exceeded? It would seem
It seems like a false positive if we warned that a 64MB image was rendered in an area that would only require a 63MB

Reply:
I was considering this, but I think the absolute number is probably more useful for a few reasons:

- The developers likely to use this are working on smaller mobile/embedded devices, and are concerned with KB usage rather than MB usage.
- A developer working on a screen with a high enough resolution to show a 64MB image (e.g. a 12MB/4K image) are probably not concerned with this anyway - a machine with that high a resolution will have enough memory where it shouldn't be a big deal.

I was considering whether we could add a percentage, but that seems to complicate the interface. That said, I'm not clear on how this would get exposed to a developer in an IDE/devtools - knowing that might help here.

Reply:
If we can only choose one approach: either multiplicative (percentage increase), or additive (absolute KB increase), I'd prefer the multiplicative too. Here's a more realistic case for a low-end device: suppose we're showing 100 thumbnails where each image's full resolution takes 100KB, but thumbnail resolution only takes 50KB. An additive threshold of 100KB would not report warning, but together (x100) they use 5MB more memory. A multiplicative 50% threshold would catch that.

Reply:
That's a really good point.

I'm trying to think of a way we could capture both of these:

- A list of small images that together add up to a lot (e.g. a list of even small images that could still be half the size they actually are).
- A single large image that, while a small percentage, still adds up to a lot (e.g. a 12MP camera image rendered even at device resolution)

We could add two properties, but that seems harder to manage and reason about to me. We could also probably track the images in a static map, and then check that the whole map doesn't exceed the maximum allowable KB.

Reply:
It may also be simpler to have two properties (or a complex property) - something for a percentage over and an absolute max. We can figure out how tooling might set that later.

Reply:
I suspect the ideal UI for this is not individual errors but a table displayed in DevTools showing the aggregate unnecessary memory usage across all images and a sorted list of the worst offenders based on a combination of byte delta and ratio of waste.

Reply:
That seems to lend towards just having a static variable that tracks this, and that we can push to devtools on demand.

I'll tinker with that some more.

Reply:
great! I think this will be a really useful debugging feature.

Reply:
So I think rather than two properties, this should work out if we use a percentage but base it on the natural log of the operands. So basically

```dart
if (math.log(displayBytes) / math.log(decodedBytes) >= allowedOverheadPercentage) {
 // bad
}
```

This should mean that you could pick a fairly reasonable percentage that will more heavily weight towards larger overheads, where it gets more important. WDYT?

Reply:
that is clever but it would be really hard for users to get a mental model of what it means. I'm not sure I have a good mental model of what value of allowedOverheadPercentage would make sense as a threshold using logs and whether small images and large images are treated in a desirable way. It would be easier to say:
we flag images as being too large if one of the following is true:
displayBytes is more than X kb smaller than decodedBytes and the percentage increase in size it at least Y
or displayByes is more than Z percent smaller than decodedBytes and the # of wasted bytes is at least Q

Reply:
The criticism is fair. I'm still a bit concerned by the complexity getting introduced here.

Here's another thought:

Rather than having any threshold for this, we just have a method exposed over a VM Service extension to dump this out. We also call `postEvent` with all the details we have when drawing an image.

In an test, you could assert that the total memory overhead is whatever you want it to be. And in devtools, we can offer some kind of visualization of this.

Reply:
`math.log(displayBytes) / math.log(decodedBytes)` is difficult for me to understand too. It's undefined if `decodedBytes == 1KB`, or negative if `displayBytes > 1KB, decodedBytes < 1KB`.

It would be easier to understand if it were `math.log(displayBytes) - math.log(decodedBytes)`, which is equivalent to `math.log(displayBytes / decodedBytes)`, so there will be a 1:1 mapping between the multiplicative threshold and this log threshold.

Reply:
yes I agree that this is the sort of threshold that is better to iterate on in devtools than to hard code in the flutter framework. Based on user studies and feedback from users we can probably come up with much better thresholds than what we could guess here. +100 to using postEvent to just dump all images added and removed from the widget tree so filtering and display can be done separately.